app-creator/client/src/widgets/template-wizard/template-wizard.ts

```diff
@@ -0,0 +1,614 @@
+          <template-card
+            id="${id}"
+            title="${name}"
+            imageUrl="${imageUrl}"
+            ?selected=${this.selectedTemplateID === id}
+            ?showTitle=${showTitle}
+            .onSelection=${() => this.handleTemplateSelection(id)}
+          ></template-card>
+        `,
+      };
+    });
+  }
+
+  /**
+   * Sets the query property when an onsearch event is dispatched from the
+   * searchbar widget.
+   */
+  handleSearch({ detail: { query } }: onSearchEvent) {
+    this.query = query.trim();
+  }
+
+  /**
+   * Creates a text input element.
+   */
+  getTextInput(
+    title: string,
+    value: string,
+    placeholder: string,
+    key: string,
+    disabled: boolean,
+    required?: boolean
+  ): TemplateResult {
+    return html`
+    <div class="config-input-container">
+        ${this.getInputHeader(title)}
+        <input
+            class='config-input text-input'
+            placeholder=${placeholder}
+            @keyup=${(e: Event) => {
+              this.config[key] = (e.target as HTMLInputElement).value;
+              this.requestUpdate();
+            }}
+            ?disabled=${disabled}
+            ?required=${required}
+            value=${value}
+        ></input>
+    </div>
+    `;
+  }
+
+  /**
+   * Creates a select input element.
+   */
+  getSelectInput(
+    key: string,
+    title: string,
+    value: string,
+    items: string[]
+  ): TemplateResult | {} {
+    if (items == null) {
+      return nothing;
+    }
+    const optionList = [];
+    for (const item of items) {
+      optionList.push(html`<option value="${item}" ?selected=${item === value}
+        >${item}</option
+      >`);
+    }
+
+    return html`
+      <div class="config-input-container config-select-input">
+        ${this.getInputHeader(title)}
+        <select
+          name="${title}"
+          class="config-input"
+          .value="${this.config[key]}"
+          @change=${(e: Event) => {
+            this.config[key] = (e.target as HTMLSelectElement).value;
+          }}
+        >
+          ${optionList}
+        </select>
+      </div>
+    `;
+  }
+
+  /**
+   * Takes in a template object and applies a selected color palette to it.
+   */
+  applyConfiguration(template: any, color: Palette) {
+    const widgets = template.widgets;
+    for (const widgetID in widgets) {
+      // Set the background color of panel elements. None panel elements start with a transparent background.
+      if (
+        widgetID.startsWith('panel') ||
+        widgetID.startsWith('button') ||
+        widgetID.startsWith('sidemenu')
+      ) {
+        widgets[widgetID].style.backgroundColor =
+          palette[color].backgroundColor;
+      }
+
+      // Apply color property on none panel elements.
+      if (!widgetID.startsWith('panel')) {
+        widgets[widgetID].style.color = palette[color].color;
+      }
+
+      // Apply map styling.
+      if (widgetID.startsWith('map')) {
```

Comment:
It looks like we're applying `color` and `backgroundColor` to `map` elements, which is undefined. Should we move this to the top of the loop and put all the other logic in the `else`?